File: scripts/cleanup-i18n.py

import re

file_path = 'client/src/contexts/i18n.tsx'

with open(file_path, 'r', encoding='utf-8') as f:
    lines = f.readlines()

current_lang = None
seen_keys = {
    'en': set(),
    'zh': set()
}
duplicate_indices = []

# Regex to match key-value pairs like: "key": "value",
# It should capture the key content inside the quotes
key_regex = re.compile(r'^\s*"([^"]+)"\s*:\s*')

for i, line in enumerate(lines):
    trimmed = line.strip()

    # Detect start of language blocks
    if trimmed == 'en: {':
        current_lang = 'en'
        continue
    if trimmed == 'zh: {':
        current_lang = 'zh'
        continue

    # Detect end of language blocks
    if (current_lang == 'en' or current_lang == 'zh') and trimmed.startswith('},'):
        current_lang = None
        continue

    if current_lang:
        # Check for comments
        if trimmed.startswith('//'):
            continue

        match = key_regex.match(line)
        if match:
            key = match.group(1)
            if key in seen_keys[current_lang]:
                print(f"Duplicate found in {current_lang}: {key} at line {i + 1}")
                duplicate_indices.append(i)
            else:
                seen_keys[current_lang].add(key)

print(f"Total duplicates found: {len(duplicate_indices)}")

if duplicate_indices:
    new_lines = [line for i, line in enumerate(lines) if i not in duplicate_indices]
    with open(file_path, 'w', encoding='utf-8') as f:
        f.writelines(new_lines)
    print('Fixed duplicates and wrote back to file.')

File: scripts/cleanup-i18n.js

const fs = require('fs');
const path = require('path');

const filePath = path.join(__dirname, '../client/src/contexts/i18n.tsx');
const content = fs.readFileSync(filePath, 'utf8');
const lines = content.split('\n');

let currentLang = null;
const seenKeys = {
  en: new Set(),
  zh: new Set()
};
const duplicateLines = [];

// Regex to match key-value pairs like: "key": "value",
const keyRegex = /^\s*"([^"]+)"\s*:\s*/;

for (let i = 0; i < lines.length; i++) {
  const line = lines[i];
  const trimmed = line.trim();

  // Detect start of language blocks
  if (trimmed === 'en: {') {
    currentLang = 'en';
    continue;
  }
  if (trimmed === 'zh: {') {
    currentLang = 'zh';
    continue;
  }

  // Detect end of language blocks
  if ((currentLang === 'en' || currentLang === 'zh') && trimmed === '},') {
    currentLang = null;
    continue;
  }

  if (currentLang) {
    // Check for comments
    if (trimmed.startsWith('//')) continue;

    const match = line.match(keyRegex);
    if (match) {
      const key = match[1];
      if (seenKeys[currentLang].has(key)) {
        console.log(`Duplicate found in ${currentLang}: ${key} at line ${i + 1}`);
        duplicateLines.push(i);
      } else {
        seenKeys[currentLang].add(key);
      }
    }
  }
}

console.log(`Total duplicates found: ${duplicateLines.length}`);

// If we want to fix it, we can filter out the duplicate lines
if (duplicateLines.length > 0) {
    const newLines = lines.filter((_, index) => !duplicateLines.includes(index));
    fs.writeFileSync(filePath, newLines.join('\n'));
    console.log('Fixed duplicates and wrote back to file.');
}
